nova/virt/vmwareapi/vm_util.py

```diff
@@ -275,20 +274,61 @@ def _vm_ref_cache(id, func, session, data):
 
 
 def vm_ref_cache_from_instance(func):
-    @functools.wraps(func)
+    @six.wraps(func)
     def wrapper(session, instance):
-        id = instance.uuid
-        return _vm_ref_cache(id, func, session, instance)
+        id_ = instance.uuid
+        return _vm_ref_cache(id_, func, session, instance)
     return wrapper
 
 
 def vm_ref_cache_from_name(func):
-    @functools.wraps(func)
+    @six.wraps(func)
     def wrapper(session, name):
-        id = name
-        return _vm_ref_cache(id, func, session, name)
+        id_ = name
+        return _vm_ref_cache(id_, func, session, name)
     return wrapper
 
+
+def vm_ref_cache_heal_from_instance(func):
+    """Decorator for a function working with a cached ManagedObject reference
+    for an instance
+
+    Invalidates the cache in case of a ManagedObjectNotFoundException
+    Most functions rely on the reference to be stable over the life-time of an
+    instance
+    However a vm-migration can move the vm out of the host, and we would end up
```

Comment:
We need to clarify "host" here. "nova-compute-*" (i.e. cluster) migration only matters, if we change vCenter. Otherwise, the moref will stay the same.

Reply:
Yes, I mean the nova-compute-host here, not the esxi-host. The mo-ref may change due to de-registering and registering an VM.

Reply:
I've updated the comment. And noted that actually the main motivation was to ensure that `get_vm_ref` will raise an `InstanceNotFound`, because `ManagedObjectNotFoundException` is rarely handled anywhere